redis/cluster.py

```diff
@@ -2397,30 +2556,242 @@ def eval(self):
+            output = redis_node.parse_response(conn, command_name, **options)
+
+        except (AskError, MovedError) as slot_error:
+            self.slot_migrating = True
+            raise slot_error
+
+        if command_name in self.UNWATCH_COMMANDS:
+            self.watching = False
+        return output
+
+    def _disconnect_reset_raise(self, conn, error):
+        """
+        Close the connection, reset watching state and
+        raise an exception if we were watching,
+        retry_on_timeout is not set,
+        or the error is not a TimeoutError
+        """
+        if not conn:
+            conn = self.transaction_connection
+
+        if conn:
+            conn.disconnect()
+
+        # if we were already watching a variable, the watch is no longer
+        # valid since this connection has died. raise a WatchError, which
+        # indicates the user should retry this transaction.
+        if self.watching:
+            self.reset()
+            raise WatchError(
+                "A ConnectionError occurred on while watching one or more keys"
+            )
+        # if retry_on_timeout is not set, or the error is not
+        # a TimeoutError, raise it
+        if not (conn and conn.retry_on_timeout and isinstance(error, TimeoutError)):
+            self.reset()
+            raise
+
+    def _reinitialize_on_error(self, error):
+        if self.watching:
+            if self.slot_migrating and self.executing:
+                raise WatchError("Slot rebalancing ocurred while watching keys")
+            if self.cluster_error:
+                raise RedisClusterException("Cluster error ocurred while watching keys")
+
+        if self.slot_migrating or self.cluster_error:
+            if self.transaction_connection:
+                self.transaction_connection = None
+
+            self.reinitialize_counter += 1
+            if self._should_reinitialized():
+                self.nodes_manager.initialize()
+                self.reinitialize_counter = 0
+            else:
+                self.nodes_manager.update_moved_exception(error)
+
+            self.slot_migrating = False
+            self.cluster_error = False
+        self.executing = False
+
+    def _execute_transaction_with_retries(
+        self, stack: List["PipelineCommand"], raise_on_error: bool
+    ):
+        retry = Retry(
+            default_backoff(),
+            self.cluster_error_retry_attempts,
+        )
+        retry.update_supported_errors([AskError, MovedError])
+        return retry.call_with_retry(
+            lambda: self._execute_transaction(stack, raise_on_error),
+            self._reinitialize_on_error,
         )
 
-    def _execute_transaction(self, *args, **kwargs):
-        """ """
-        raise RedisClusterException("method _execute_transaction() is not implemented")
+    def _execute_transaction(
+        self, stack: List["PipelineCommand"], raise_on_error: bool
+    ):
+        if len(self.pipeline_slots) > 1:
```

Comment:
As I mentioned above looks confusing, I think it makes sense to keep a transaction slot as a separate variable